### PR TITLE
Allow VM boot modes to change a VM's default user

### PR DESCRIPTION
**Pull request by ArrayBolt3**
Pretty straightforward, adds another feature category `boot-mode.default-user.<boot-mode-name>`. Regression tests pass on my end, and the feature itself works in testing. I'm able to remove the qrexec overrides for Whonix-Workstation, add a default-user feature for sysmaint mode, and have sysmaint mode appear to function properly. Software updates also work again as dom0 can now run commands in VMs as root again.

### CODE REVIEW

**Comment by ArrayBolt3** [comment]
The one line missing coverage appears to be a change of a `self.default_user` reference to a `self.get_default_user()` call. The original line wasn't covered either.

File: qubes/ext/core_features.py [hunk 1]
```diff
@@ -136,19 +138,22 @@ async def qubes_features_request(self, vm, event, untrusted_features):
...
+            bootmode_name = bootmode_key_parts[2]
+            if bootmode_name == "":
+                continue
+            if (
+                f"boot-mode.kernelopts.{bootmode_name}" not in new_bootmode_info
+            ) and bootmode_name != "default":
+                continue
+            if untrusted_feature_key.startswith(
+                "boot-mode.name."
+            ) or untrusted_feature_key.startswith("boot-mode.default-user."):
```
Comment (marmarek) [inline qubes/ext/core_features.py:156]: You allow setting `boot-mode.default-user.default`, but it's later ignored in `get_default_user()`. Please make it consistent (if you have some good argument for it, I can be convinced to allowing default user for the default boot mode too; otherwise don't allow setting it here).

File: qubes/tests/vm/qubesvm.py
Comment (marmarek) [inline qubes/tests/vm/qubesvm.py:1]: Maybe also worth extending integration test in qubes/tests/integ/basic.py, to check if really new default user got used?

**Review by marmarek — CHANGES_REQUESTED** [review]

**Comment by marmarek** [comment]
Besides the comments above, it looks fine and together with whonix developers repo updates seems to fix the issue.

**Comment by ArrayBolt3** [comment]
This hasn't fallen off my radar, I've just been having to tackle other work responsibilities and have had very limited time (too limited to look at this) for a few days. I'm scheduled to resume work on Whonix and Kicksecure again starting today, and this is my top priority to look at, so expect something later today.

File: qubes/ext/core_features.py (hunk 1 above)
Comment (ArrayBolt3) [reply qubes/ext/core_features.py]: Fixed, test added for this edge case.

File: qubes/tests/vm/qubesvm.py
Comment (ArrayBolt3) [reply qubes/tests/vm/qubesvm.py:1]: Good idea, implemented.

**Comment by ArrayBolt3** [comment]
@marmarek Should be ready for re-review.

File: qubes/tests/integ/basic.py [hunk 2]
```diff
@@ -513,6 +513,32 @@ async def _test_bootmode(self, tpl, vm):
...
+        )
+        await vm.run_for_stdio(
+            "chmod +x " "/etc/qubes/post-install.d/50-test.sh", user="root"
+        )
+        await vm.run_service_for_stdio("qubes.PostInstall", user="root")
+        await vm.shutdown(wait=True)
+
+        await vm.start()
+        user_id = (await vm.run_for_stdio("id -un"))[0].decode()
+        self.assertEquals(user_id, "altuser")
```
[the remark this answers is not in the document]
Comment (ArrayBolt3) [reply qubes/tests/integ/basic.py]: Ah, that's because the tests aren't being run under dom0. That's going to be a bit trickier to do, I'll try to run them under dom0 locally though.

[the remark this answers is not in the document]
Comment (ArrayBolt3) [reply qubes/tests/integ/basic.py]: Kk. I should still have a dom0 testing environment for in the future though, so my time setting it up won't be wasted, but I will go ahead and let openQA handle it for this time around. Thank you!

**Review by marmarek — APPROVED** [review]

**Comment by marmarek** [comment]
@ArrayBolt3 do you remember why it isn't plugged into `default_user` property, but a bunch of specific places instead of using that property directly? I think there was some reason for it, just can't find it now...

**Comment by ArrayBolt3** [comment]
Unfortunately I don't remember anymore. I searched Matrix chat history and found some context, but not enough to fully piece together the original rationale.
